### PR TITLE
Refactor stack IR / binary writer (NFC)

### PR DESCRIPTION
**Pull request by aheejin**
Previously `StackWriter` and its subclasses had routines for all three
modes (`Binaryen2Binary`, `Binaryen2Stack`, and `Stack2Binary`) within a
single class. This splits routines for each in a separate class and
also factors out binary writing into a separate class
(`BinaryInstWriter`) so other classes can make use of it.

The new classes are:
- `BinaryInstWriter`:
   Binary instruction writer. Only responsible for emitting binary
   contents and no other logic
- `BinaryenIRWriter`: Converts binaryen IR into something else
  - `BinaryenIRToBinaryWriter`: Writes binaryen IR to binary
  - `StackIRGenerator`: Converts binaryen IR to stack IR
- `StackIRToBinaryWriter`: Writes stack IR to binary

I didn't use the existing `***StackWriter` names because I couldn't
exactly understand their meanings, but I can rename classes to resemble
previous names if people suggest it.

### CODE REVIEW

File: src/wasm-stack.h [hunk 1]
```diff
@@ -162,2152 +121,648 @@ class StackWriter : public OverriddenVisitor<StackWriter<Mode, Parent>> {
...
-  StackInst* makeStackInst(Expression* origin) {
-    return makeStackInst(StackInst::Basic, origin);
-  }
 };
 
-// Write out a single expression, such as an offset for a global segment.
-template<typename Parent>
-class ExpressionStackWriter
-  : StackWriter<StackWriterMode::Binaryen2Binary, Parent> {
+// Takes a binaryen IR and converts it to something else (binary or stack IR)
```
Comment (kripken) [inline src/wasm-stack.h]: `Takes a` => `Takes`

File: src/wasm/wasm-binary.cpp [hunk 2]
```diff
@@ -295,7 +295,8 @@ void WasmBinaryWriter::writeFunctionSignatures() {
 }
 
 void WasmBinaryWriter::writeExpression(Expression* curr) {
-  ExpressionStackWriter<WasmBinaryWriter>(curr, *this, o, debug);
+  BinaryenIRToBinaryWriter writer(*this, o);
+  writer.visit(curr);
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: this and similar things can be `BinaryenIRToBinaryWriter(*this, o).visit(curr);` I think, on one line?

File: src/wasm-stack.h (hunk 1 above)
Comment (aheejin) [reply src/wasm-stack.h]: Done

File: src/wasm/wasm-binary.cpp (hunk 2 above)
Comment (aheejin) [reply src/wasm/wasm-binary.cpp]: Done. Also fixed a few more similar places.

**Comment by kripken** [comment]
Looks like this CI problem is unrelated to your PR, I just saw it after I merged another green PR to master. It was green 2 days ago, but was red when it landed, so some CI infra change over the last day or so seems to be at fault,

https://travis-ci.org/WebAssembly/binaryen/jobs/562365712

File: src/wasm-stack.h [hunk 3]
```diff
@@ -162,2152 +121,648 @@ class StackWriter : public OverriddenVisitor<StackWriter<Mode, Parent>> {
...
-      // Finish the later parts of all the parent blocks.
-      while (!parents.empty()) {
-        auto* parent = parents.back();
-        parents.pop_back();
-        visitChildren(parent, 1);
-        afterChildren(parent);
-      }
-      return;
+  // avoiding heavy recursion. We only start to do this if we see it will help
+  // us (to avoid allocation of the vector). Note that Stack2Binary mode we
```
Comment (tlively) [inline src/wasm-stack.h]: Stale comment here about `Stack2Binary` mode.

File: src/wasm/wasm-stack.cpp [hunk 4]
```diff
@@ -0,0 +1,1591 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
```
Comment (tlively) [inline src/wasm/wasm-stack.cpp]: 2019

File: src/wasm-stack.h (hunk 3 above)
Comment (aheejin) [reply src/wasm-stack.h]: Deleted.

File: src/wasm/wasm-stack.cpp (hunk 4 above)
Comment (aheejin) [reply src/wasm/wasm-stack.cpp]: Done

**Pull request merged**